File: python/linkedlist/reverse.py
class Node():
    def __init__(self, val=None):
        self.val = val
        self.next = None


def arr2linked(arr):
    res = cur = Node(arr[0])
    for a in arr[1:]:
        cur.next = Node(a)
        cur = cur.next

    return res


def link2arr(link):
    list = []
    while link:
        list.append(link.val)
        link = link.next

    return list


def reverse(head):
    prev = None
    while head:
        cur = head
        head = head.next
        cur.next = prev
        prev = cur

    return prev


if __name__ == "__main__":
    list = [1, 2, 3, 4, 5, 6]
    head = arr2linked(list)
    head = reverse(head)

    assert link2arr(head) == [6, 5, 4, 3, 2, 1]

File: cpp/linked-list/reverse.cpp
/**
Reverse a linked list from position m to n. Do it in one-pass.

Note: 1 ≤ m ≤ n ≤ length of list.

Example:

Input: 1->2->3->4->5->NULL, m = 2, n = 4
Output: 1->4->3->2->5->NULL
 * **/

struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(0) {}
};

class Solution
{
  public:
    ListNode *reverseBetween(ListNode *head, int m, int n)
    {
        ListNode *res = new ListNode(0);
        res->next = head;
        ListNode *prev = res;
        ListNode *next = 0;

        for (int i = 1; i < m; i++)
            prev = prev->next;
        ListNode *cur = prev->next;
        for (int i = 0; i < n - m; i++)
        {
            next = cur->next;
            cur->next = next->next;
            next->next = prev->next;
            prev->next = next;
        }
        return res->next;
    }
};